File: java/CavanJava/src/com/cavan/java/CavanWaveArray.java
package com.cavan.java;

@SuppressWarnings("serial")
public class CavanWaveArray extends CavanTimedArray<Double> {

	private boolean mNeedFind = true;
	protected CavanTimedNode<Double> mMinNode;
	protected CavanTimedNode<Double> mMaxNode;

	public CavanWaveArray(long overtime) {
		super(overtime);
	}

	protected void updateExtremeNode() {
		mMinNode = mMaxNode = get(0);

		for (int i = size() - 1; i > 0; i--) {
			CavanTimedNode<Double> node = get(i);

			if (node.getValue() > mMaxNode.getValue()) {
				mMaxNode = node;
			} else if (node.getValue() < mMinNode.getValue()) {
				mMinNode = node;
			}
		}
	}

	protected void setMaxNode(CavanTimedNode<Double> node) {
		mMaxNode = node;
	}

	protected void setMinNode(CavanTimedNode<Double> node) {
		mMinNode = node;
	}

	public CavanTimedNode<Double> getMaxNode() {
		return mMaxNode;
	}

	public CavanTimedNode<Double> getMinNode() {
		return mMinNode;
	}

	public double getMaxValue() {
		if (mMaxNode == null) {
			return 0;
		}

		return mMaxNode.getValue();
	}

	public double getMinValue() {
		if (mMinNode == null) {
			return 0;
		}

		return mMinNode.getValue();
	}

	@Override
	protected CavanTimedNode<Double> removeTimedNode(int index) {
		CavanTimedNode<Double> node = super.removeTimedNode(index);
		if (node == mMinNode || node == mMaxNode) {
			mNeedFind = true;
		}

		return node;
	}

	public void addWaveNode(CavanTimedNode<Double> node) {
		super.addTimedNode(node);

		if (mNeedFind) {
			mNeedFind = false;
			updateExtremeNode();
		} else if (node.getValue() > mMaxNode.getValue()) {
			setMaxNode(node);
		} else if (node.getValue() < mMinNode.getValue()) {
			setMinNode(node);
		}
	}

	public void addWaveValue(double value) {
		addWaveNode(new CavanTimedNode<Double>(value));
	}
}

File: xcode/JwaooToy/JwaooToy/CavanWaveArray.m
//
//  CavanWaveArray.m
//  JwaooToy
//
//  Created by 曹福昂 on 16/8/31.
//  Copyright © 2016年 曹福昂. All rights reserved.
//

#import "CavanWaveArray.h"

@implementation CavanWaveNode

@synthesize value = mValue;

- (CavanWaveNode *)initByValue:(double)value {
    if (self = [super init]) {
        mValue = value;
    }

    return self;
}

@end

@implementation CavanWaveArray

- (CavanWaveArray *)initByOvertime:(NSTimeInterval)overtime {
    if (self = [super initByOvertime:overtime]) {
        mNeedFind = true;
    }

    return self;
}

- (void)removeTimedNode:(NSInteger)index {
    CavanTimedNode *node = [mNodes objectAtIndex:index];
    if (node == mMaxNode || node == mMinNode) {
        mNeedFind = true;
    }

    [super removeTimedNode:index];
}

- (void)onExtremeUpdated {
}

- (void)updateExtremeNode {
    mMinNode = mMaxNode = mNodes.firstObject;

    for (NSInteger index = mNodes.count - 1; index > 0; index--) {
        CavanWaveNode *node = [mNodes objectAtIndex:index];
        if (node.value > mMaxNode.value) {
            mMaxNode = node;
        } else if (node.value < mMinNode.value) {
            mMinNode = node;
        }
    }
}

- (void)putWaveValue:(double)value {
    CavanWaveNode *node = [[CavanWaveNode alloc] initByValue:value];

    [self addTimedNode:node];

    if (mNeedFind) {
        mNeedFind = false;
        [self updateExtremeNode];
    } else if (value > mMaxNode.value) {
        mMinNode = node;
    } else if (value < mMinNode.value) {
        mMaxNode = node;
    } else {
        return;
    }

    [self onExtremeUpdated];
}

@end
